### PR TITLE
fix: support metric collections for autoscaling group

### LINKED ISSUE #6453: Unable to switch on metric collections for autoscaling group
1. ASG has a property `MetricsCollection`. It is off by default.
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-group.html#cfn-autoscaling-autoscalinggroup-autoscalinggroupname

2. CfnAutoScalingGroupProps supports this property

3. AutoScalingGroup class do not have any means to set this property up
https://github.com/aws/aws-cdk/blob/v1.25.0/packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts#L495

### Environment

  - **CLI Version      :** 1.25.0
  - **Framework Version:** 
  - **OS               :** MacOs
  - **Language         :** TypeScript

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by fogfish**
fixes #6453

### CODE REVIEW

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 1]
```diff
@@ -21,6 +21,58 @@ import { BlockDevice, EbsDeviceVolumeType } from './volume';
...
+  /** GroupStandbyInstances */
+  STANDBY_INSTANCES = "GroupStandbyInstances",
+
+  /** GroupTerminatingInstances */
+  TERMINATING_INSTANCES = "GroupTerminatingInstances",
+
+  /** GroupTotalInstances */
+  TOTAL_INSTANCES = "GroupTotalInstances",
+}
+
```
[the remark this answers is not in the document]
Comment (fogfish) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: I would love to remove these comments but the build system gonna give me a error that documentation is required for exported types.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 2]
```diff
@@ -21,6 +21,58 @@ import { BlockDevice, EbsDeviceVolumeType } from './volume';
...
+/**
+ * The frequency at which Amazon EC2 Auto Scaling sends aggregated data to CloudWatch.
+ */
+export type MetricGranularity = "1Minute";
+
+/**
+ * Auto Scaling group metrics
+ */
+export enum GroupMetric {
+  /** GroupMinSize */
```
[the remark this answers is not in the document]
Comment (fogfish) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: - [ ] Change to `MetricGroup`

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 3]
```diff
@@ -21,6 +21,58 @@ import { BlockDevice, EbsDeviceVolumeType } from './volume';
...
 const NAME_TAG: string = 'Name';
 
+/**
+ * Describes the group metrics that an Amazon EC2 Auto Scaling group sends to Amazon CloudWatch.
+ * These metrics describe the group rather than any of its instances.
+ * For more information, see  Monitoring Your Auto Scaling Groups and Instances Using Amazon CloudWatch in the Amazon EC2 Auto Scaling User Guide.
+ * https://docs.aws.amazon.com/autoscaling/ec2/userguide/as-instance-monitoring.html
+ */
+export interface AutoScalingGroupMetric {
+  /**
```
[the remark this answers is not in the document]
Comment (fogfish) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: - [ ] Change to `MetricsCollection`

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 4]
```diff
@@ -21,6 +21,58 @@ import { BlockDevice, EbsDeviceVolumeType } from './volume';
  */
 const NAME_TAG: string = 'Name';
 
+/**
+ * Describes the group metrics that an Amazon EC2 Auto Scaling group sends to Amazon CloudWatch.
+ * These metrics describe the group rather than any of its instances.
+ * For more information, see  Monitoring Your Auto Scaling Groups and Instances Using Amazon CloudWatch in the Amazon EC2 Auto Scaling User Guide.
+ * https://docs.aws.amazon.com/autoscaling/ec2/userguide/as-instance-monitoring.html
+ */
```
[the remark this answers is not in the document]
Comment (fogfish) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: - [ ] Update the link in the doc

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 5]
```diff
@@ -21,6 +21,58 @@ import { BlockDevice, EbsDeviceVolumeType } from './volume';
...
+   * The list of Auto Scaling group metrics to collect. If you specify Granularity and don't specify any metrics, all metrics are enabled.
+   */
+  readonly metrics: GroupMetric[],
+}
+
+/**
+ * The frequency at which Amazon EC2 Auto Scaling sends aggregated data to CloudWatch.
+ */
+export type MetricGranularity = "1Minute";
+
```
[the remark this answers is not in the document]
Comment (fogfish) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: You build system gives an error if I create enum with one entry. This is exactly how I've implemented at first time.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 6]
```diff
@@ -21,6 +21,58 @@ import { BlockDevice, EbsDeviceVolumeType } from './volume';
...
+ */
+export interface AutoScalingGroupMetric {
+  /**
+   * The frequency at which Amazon EC2 Auto Scaling sends aggregated data to CloudWatch.
+   */
+  readonly granularity: MetricGranularity,
+
+  /**
+   * The list of Auto Scaling group metrics to collect. If you specify Granularity and don't specify any metrics, all metrics are enabled.
+   */
```
[the remark this answers is not in the document]
Comment (fogfish) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: - [ ] fix typo

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 7]
```diff
@@ -171,6 +223,13 @@ export interface CommonAutoScalingGroupProps {
...
    */
   readonly healthCheck?: HealthCheck;
+
+  /**
+   * Enables the monitoring of group metrics of an Auto Scaling group.
+   *
+   * @default - metrics collection is disabled.
+   */
+  readonly metricsCollection?: AutoScalingGroupMetric[]
 }
```
[the remark this answers is not in the document]
Comment (fogfish) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: It is called metricCollection at CF.
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-group.html#cfn-as-group-metricscollection

**Comment by fogfish** [comment]
Thank you very much for your suggestion about the union pattern. I think it is helpful in may use-cases. Unfortunately, I've confused about its usability from developers perspective in this particular use-case. Let's step into developer shoe:

As a developer I want to enable ASG monitoring so that ...

As a developer I want to enable monitoring of specific metrics so that ...

It became more friendly for developer to use boolean flag just to enable monitoring. Therefore I would tune ASG properties to accept Sum types

```ts
class ASGProps {
  readonly enableMonitoring: boolean | MetricsCollection[]
}


new autoscaling.AutoScalingGroup(stack, 'MyFleet', {
   enableMonitoring: true
}

new autoscaling.AutoScalingGroup(stack, 'MyFleet', {
  enableMonitoring: [
     {
       granularity: Granularity.ONE_MINUTE,
       metrics: [MetricsGroup. GROUP_IN_SERVICE_INSTANCES]
     },
  ]
}
```

**Comment by fogfish** [comment]
@NetaNir Any thoughts? I would not make any change to the PR until we agree about API.

**Comment by NetaNir** [comment]
Hi @fogfish!

The single enum bug was fixed! 

As for the API, the experience in Java, C# and typically other statically typed languages without type unions will be inadequate, which is why we try to avoid it in the CDK (we currently can't prevent it in jsii since it will be a breaking change) and use the union like pattern.

**Comment by huangzbaws** [comment]
Is there any update for this issue? My project is still waiting for this feature to be closed. Thanks!

**Comment by NetaNir** [comment]
Closing this PR in favor of https://github.com/aws/aws-cdk/pull/7432

**Pull request closed without merging**